### PR TITLE
Add Phase model

### PR DESCRIPTION
**Pull request by hancush**
## Overview

This PR adds a `Phase` model. A phase is a sub-project unit of work.

This PR intentionally omits material changes to the Project creation / editing interfaces. Will do in #80. It does update some view code so the related Phase and PhaseFinances objects are created when a project is added, but this implementation should not be considered final. I've left TODOs in places I've commented or otherwise made stopgap changes to the existing code.

Related to but doesn't quite close (?) #77.

## Testing Instructions

* This PR provides tests for the new data migration, both forwards and backwards. Review those tests to confirm I've implemented the data migration as we like, then confirm CI passes.

### CODE REVIEW

**Comment by hancush** [comment]
@fgregg Before I go further on this, probably a good idea to get your input. Let me know what you think in a review.

**Comment by fgregg** [comment]
@hancush this is looking like it's going in the right direction!

File: asset_dashboard/migrations/0019_add_task.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
+# Generated by Django 3.2.9 on 2021-12-06 21:06
```
Comment (hancush) [inline asset_dashboard/migrations/0019_add_task.py:1]: This migration is pretty involved and needs a close review!

File: asset_dashboard/migrations/0019_add_task.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        ),
```
Comment (hancush) [inline asset_dashboard/migrations/0019_add_task.py:235]: Django wanted to add these after I pinned the `dj-money` requirement.

File: asset_dashboard/models.py [hunk 3]
```diff
@@ -37,29 +37,41 @@ class Project(models.Model):
...
+        return self.name
+
+
+class Task(models.Model):
+    """
+    A sub-project unit of work. Projects without defined tasks are assigned a
+    default Task of type "implementation".
+    """
+
+    # TODO: Identify better choice list (e.g., what's the difference between
```
Comment (hancush) [inline asset_dashboard/models.py:50]: We'll need to rope Garret in for this, I think. Can address separately.

File: requirements.txt [hunk 4]
```diff
@@ -1,21 +1,17 @@
...
-Django
-psycopg2-binary
-gunicorn
-dj-database-url
-whitenoise
-django-compressor
-sentry-sdk
-django-money
-django-datatables-view
+Django==3.1.7
```
Comment (hancush) [inline requirements.txt:1]: I pinned everything here by opening a console against the production app and running `pip freeze` to generate the current versions of our dependencies on the live site.

File: tests/test_migrations.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
+import random
+
+import pytest
+
+
```
Comment (hancush) [inline tests/test_migrations.py:5]: See https://github.com/wemake-services/django-test-migrations for more on this neat plug-in, `django-test-migrations`!

File: requirements.txt [hunk 6]
```diff
@@ -1,21 +1,17 @@
...
 # Testing requirements
-pytest
-pytest-django
-pytest-mock
-flake8
-
-# Project setup requirements
-pip-chill
+pytest==6.2.2
+pytest-django==4.5.1
```
Comment (hancush) [inline requirements.txt:14]: The only exception is I upgraded `pytest-django` for multi-DB support.

**Comment by hancush** [comment]
Note to self: Update fixtures. (@fgregg This should not preclude review.)

File: asset_dashboard/models.py [hunk 7]
```diff
@@ -37,29 +37,41 @@ class Project(models.Model):
...
     ]
 
+    sequence = models.IntegerField(default=1)
+    project = models.ForeignKey('Project', on_delete=models.CASCADE, related_name='tasks')
+    task_type = models.TextField(choices=TASK_TYPE_CHOICES, null=True, blank=True)
     estimated_bid_quarter = models.TextField(choices=BID_QUARTER_CHOICES, null=True, blank=True)
-
-    def __str__(self):
-        return self.name
+    # funding = models.ForeignKey('FundingSource', on_delete=models.SET_NULL)
```
Comment (fgregg) [inline asset_dashboard/models.py:74]: what's up with this commented line?

**Review by fgregg — APPROVED** [review]
One question about a stray commented line. Otherwise this all looks good to me.

**Comment by fgregg** [comment]
@smcalilly I want you to take a look at this PR. it will impact on your work as geographic assets will be attached to Tasks. Just keep working in parallel for now, but once this is brought in, we'll have to merge the streams.

**Comment by smcalilly** [comment]
> @smcalilly I want you to take a look at this PR. it will impact on your work as geographic assets will be attached to Tasks. Just keep working in parallel for now, but once this is brought in, we'll have to merge the streams.

looks great at a first pass. makes sense to attach the assets to the tasks. all the questions i had were answered by a closer look at the PR.

File: asset_dashboard/models.py (hunk 3 above)
Comment (fgregg) [reply asset_dashboard/models.py:50]: okay we are going to go with 

Feasibilty
Design
Engineering
Construction,

Implementation,

we are going to do away with reference to phase here.

File: asset_dashboard/models.py [hunk 8]
```diff
@@ -37,29 +37,41 @@ class Project(models.Model):
...
         ('phase_1', 'Phase 1'),
         ('phase_2', 'Phase 2'),
         ('phase_3_engineering', 'Phase 3 Engineering'),
         ('phase_3_construction', 'Phase 3 Construction'),
         ('feasibility', 'Feasibility'),
         ('design', 'Design'),
-        ('construction', 'Construction')
+        ('construction', 'Construction'),
+        ('implementation', 'Implementation'),
     ]
```
Comment (fgregg) [inline asset_dashboard/models.py:61]: also need a status field

[done, unscheduled, in-progress]

**Comment by fgregg** [comment]
Good feedback from @glw! He really likes this modeling approach.

Couple high level notes:

1. @hancush is, of course right, and the name that makes sense for Garret is "phase" not "task" let's adjust the names accordingly
2. We need to add a status field (notes in line)
3. We will simply the TASK_TYPE_CHOICES, notes inline

File: asset_dashboard/models.py (hunk 7 above)
Comment (hancush) [reply asset_dashboard/models.py]: Removed.

File: asset_dashboard/models.py (hunk 8 above)
Comment (hancush) [reply asset_dashboard/models.py:56]: Added.

**Comment by hancush** [comment]
Contrary to the status indicator, review app successfully deployed at: https://ccfp-asset-d-hcg-bundle-eio7w0.herokuapp.com/

File: README.md [hunk 9]
```diff
@@ -91,10 +91,13 @@ docker-compose -f docker-compose.yml -f tests/docker-compose.yml run --rm app
...
-docker-compose run --rm app python manage.py dumpdata --natural-foreign --indent 2 \
--e contenttypes \
--e sessions \
-asset_dashboard auth > asset_dashboard/fixtures/data.json
+docker-compose run --rm app python manage.py dumpdata \
+    --natural-foreign \
+    --indent 2 \
+    -e contenttypes \
+    -e sessions \
+    -o asset_dashboard/fixtures/data.json \
```
Comment (hancush) [inline README.md:99]: You have to specify output file as a flag. Redirecting includes a bunch of extra junk (Django output) at the top of the file.

File: asset_dashboard/migrations/0019_add_phase.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
+# Generated by Django 3.2.9 on 2021-12-06 21:06
+
+from django.db import migrations, models
+import django.db.models.deletion
+
+import djmoney
+
+
+def project_to_phase(apps, schema_editor):
```
Comment (fgregg) [inline asset_dashboard/migrations/0019_add_phase.py:9]: very nice!

**Review by fgregg — APPROVED** [review]

**Pull request merged**